sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/TestGcpIoPipelineOptionsRegistrar.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp;
+
+import com.google.auto.service.AutoService;
+import org.apache.beam.sdk.annotations.Internal;
+import org.apache.beam.sdk.io.gcp.bigquery.TestBigQueryOptions;
+import org.apache.beam.sdk.io.gcp.bigtable.changestreams.BigtableChangeStreamTestOptions;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.options.PipelineOptionsRegistrar;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.ImmutableList;
+
+@AutoService(PipelineOptionsRegistrar.class)
+@Internal
+public class TestGcpIoPipelineOptionsRegistrar implements PipelineOptionsRegistrar {
+  @Override
+  public Iterable<Class<? extends PipelineOptions>> getPipelineOptions() {
+    return ImmutableList.<Class<? extends PipelineOptions>>builder()
+        .add(TestBigQueryOptions.class)
```

Comment:
These are options for `TestBigQuery` which is intended for users. Anything intended for users should be under `src/main`

Reply:
TestBigQueryOptions is still in main. The problem here is there is an AutoService to register TestBigQueryOptions, then whenever a PipelineOptionFactory.fromArgs is called, the PipelineOptionFactory tries to do some reflection access to TestBigQueryOptions which causes `NoClassDefFoundError: org/hamcrest/Matcher`. This is because TestPipelineOptions has a method return a Matcher.

This makes org.hamcrest effectively a required dependency (instead of optional one) for Beam io-gcp.

If do not register TestBigQueryOptions, users can still use it. They just need to register the option before they call PipelineOptionFactory.withValidation().

Reply:
An alternate fix will just be make it a implementation dependency here:

Reply:
Yes, making it an implementation dependency seems correct. It is, after all, an implementation dependency :-)

Reply:
hamcrest is a tiny tiny library that is super stable and has no transitive deps and not going to have security concerns - there is no reason to avoid it

Reply:
thanks, opened #30635 and close this PR